### PR TITLE
stop performFetch by a function

### PR DESCRIPTION
**Pull request by r10s**
still targets 0xdead10cc, #1202

i did not create an observer yet as this would mean additional complexity. however, when we need it because of #1157, all these changes will already help.

we need to take care that `performFetch()` always calls the `completionHandler`, even if `performFetch()` is called concurrently. this is mainly achieved by the time-check, allowing this function only once per minute. otherwise, things would be even more complicated.

### CODE REVIEW

File: deltachat-ios/AppDelegate.swift [hunk 1]
```diff
@@ -386,25 +389,41 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UNUserNotificationCenterD
...
-            }
 
-            // to avoid 0xdead10cc exceptions, scheduled jobs need to be done before we get suspended;
-            // we increase the probabilty that this happens by waiting a moment before calling completionHandler()
-            DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
-                logger.info("⬅️ fetch done")
-                completionHandler(.newData)
+            logger.info("⬅️ fetch done")
+            completionHandler(.newData)
+            self.unregisterFetchBackgroundTask()
```
Comment (cyBerta) [inline deltachat-ios/AppDelegate.swift:419]: maybe we should move that line before the completion handler is executed in order to be sure this is called before the suspension of the app happened.

File: deltachat-ios/AppDelegate.swift [hunk 2]
```diff
@@ -372,12 +376,11 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UNUserNotificationCenterD
         bgIoTimestamp = nowTimestamp
 
         // make sure to balance each call to `beginBackgroundTask` with `endBackgroundTask`
-        let backgroundTask = UIApplication.shared.beginBackgroundTask {
-            // TODO: currently, this should be okay, we are not using too much background time,
-            // so that handler should never be called.
-            // the plan is to listen to an event as DC_EVENT_ENTER_IDLE and stop fetch from there.
-            // if we have such an event, we could imprive this handler and fire the event.
-            logger.info("fetch background task will end soon")
+        fetchBackgroundTaskEnding = false
```
Comment (cyBerta) [inline deltachat-ios/AppDelegate.swift:379]: this is quite confusing to read. Is it possible to move that line to the end of `endFetch()`?

**Comment by cyBerta** [comment]
For the naming of the background task, maybe we can find sth. more descriptive. Afterburner reminds me of other things than what is meant here 🌶️🌶️🌶️

File: deltachat-ios/AppDelegate.swift (hunk 1 above)
Comment (r10s) [reply deltachat-ios/AppDelegate.swift:419]: i think, this is fine, the backgroundTask avoids the suspension.

**Comment by r10s** [comment]
please do not merge in, i have an idea about how to do this whole stuff more elegant with less "globals", have to think about that a bit more, however :)

**Comment by r10s** [comment]
i crated an alternative in https://github.com/deltachat/deltachat-ios/pull/1221

**Comment by r10s** [comment]
this pr, although having more complexity, as the same issue as #1221 - it does not call endBackgroundTask from within the handler but a moment later.

**Comment by r10s** [comment]
closing in favor to https://github.com/deltachat/deltachat-ios/pull/1223

**Pull request closed without merging**